lib/shape/data_visitor.rb

```diff
@@ -8,13 +8,15 @@ def visit(visitor = lambda {|x| x})
   protected
 
     def properties_to_visit
+      all_properties.select { |_, property| visitable?(property) }
+    end
+
+    def all_properties
       self.class.properties.merge(self.properties)
     end
 
-    def data_visitor(properties = self.properties_to_visit, visitor = lambda {|x| x})
+    def data_visitor(properties, visitor)
       properties.each_with_object({}) do |(name, property), obj|
-        next unless visitable?(property)
-
         if property.properties.present?
           obj[name] = self.data_visitor(property.properties, visitor)
```

Comment:
I think the problem with the select (as currently implemented) would show up right here :point_up: - essentially, only top-level visitables would work. We aren't ensuring that visitable is called as we recurse nested properties.